Merge type increases type instantiations count drastically

I ran into `error TS2589: Type instantiation is excessively deep and possibly infinite.` and get this result with `--extendedDiagnostics`:

```
Files:                        1527
Lines of Library:            28535
Lines of Definitions:        76407
Lines of TypeScript:         17557
Lines of JavaScript:        195957
Lines of JSON:                   0
Lines of Other:                  0
Nodes of Library:           117553
Nodes of Definitions:       273566
Nodes of TypeScript:         72611
Nodes of JavaScript:        825495
Nodes of JSON:                   0
Nodes of Other:                  0
Identifiers:                431173
Symbols:                    410281
Types:                       90155
Instantiations:            6707979            <========== the instantiations count exceeds the limitation
Memory used:               531849K
Assignability cache size:    53667
Identity cache size:          1231
Subtype cache size:           1904
Strict subtype cache size:    8066
```
tsc has a [5 million limitation on instantiation counts](https://github.com/microsoft/TypeScript/pull/32079#discussion_r297478404)

After I replace all the `Merge` with simple intersection with `Omit`, I get this:

```
Files:                        1527
Lines of Library:            28535
Lines of Definitions:        76407
Lines of TypeScript:         17522
Lines of JavaScript:        195957
Lines of JSON:                   0
Lines of Other:                  0
Nodes of Library:           117553
Nodes of Definitions:       273566
Nodes of TypeScript:         72570
Nodes of JavaScript:        825495
Nodes of JSON:                   0
Nodes of Other:                  0
Identifiers:                431152
Symbols:                    448585
Types:                       80108
Instantiations:             514523         <========== 10x smaller
Memory used:               556046K
Assignability cache size:    77690
Identity cache size:          1259
Subtype cache size:           2096
Strict subtype cache size:    8430
```

Using type-fest@2.3.4

--- a/source/merge.d.ts
+++ b/source/merge.d.ts
@@ -1,7 +1,6 @@
-import type {Except} from './except';
+import type {OmitIndexSignature} from './omit-index-signature';
+import type {PickIndexSignature} from './pick-index-signature';
 import type {Simplify} from './simplify';
-
-type Merge_<FirstType, SecondType> = Except<FirstType, Extract<keyof FirstType, keyof SecondType>> & SecondType;
 
 /**
 Merge two types into a new type. Keys of the second type overrides keys of the first type.
@@ -10,18 +9,37 @@ Merge two types into a new type. Keys of the second type overrides keys of the f
 ```
 import type {Merge} from 'type-fest';
 
-type Foo = {
-	a: number;
-	b: string;
-};
+interface Foo {
+	[x: string]: unknown;
+	[x: number]: unknown;
+	foo: string;
+	bar: symbol;
+}
 
 type Bar = {
-	b: number;
+	[x: number]: number;
+	[x: symbol]: unknown;
+	bar: Date;
+	baz: boolean;
 };
 
-const ab: Merge<Foo, Bar> = {a: 1, b: 2};
+export type FooBar = Merge<Foo, Bar>;
+// => {
+// 	[x: string]: unknown;
+// 	[x: number]: number;
+// 	[x: symbol]: unknown;
+// 	foo: string;
+// 	bar: Date;
+// 	baz: boolean;
+// }
 ```
 
 @category Object
 */
-export type Merge<FirstType, SecondType> = Simplify<Merge_<FirstType, SecondType>>;
+export type Merge<Destination, Source> = Simplify<{
+	[Key in keyof OmitIndexSignature<Destination & Source>]: Key extends keyof Source
+		? Source[Key]
+		: Key extends keyof Destination
+			? Destination[Key]
+			: never;
+} & PickIndexSignature<Destination & Source>>;

--- a/test-d/merge.ts
+++ b/test-d/merge.ts
@@ -1,4 +1,4 @@
-import {expectAssignable} from 'tsd';
+import {expectAssignable, expectError} from 'tsd';
 import type {Merge} from '../index';
 
 type Foo = {
@@ -12,3 +12,47 @@ type Bar = {
 
 const ab: Merge<Foo, Bar> = {a: 1, b: 2};
 expectAssignable<{a: number; b: number}>(ab);
+
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+interface FooInterface {
+	[x: string]: unknown;
+	[x: number]: unknown;
+	foo: string;
+	bar: symbol;
+}
+
+type BarType = {
+	[x: number]: number;
+	[x: symbol]: boolean;
+	bar: Date;
+	baz: boolean;
+};
+
+type FooBar = Merge<FooInterface, BarType>;
+
+const fooBar: FooBar = {
+	'foo-string': 'foo',
+	42: 24,
+	[Symbol(42)]: true,
+	foo: 'foo',
+	bar: new Date(),
+	baz: true,
+};
+
+expectAssignable<{
+	[x: string]: unknown;
+	[x: number]: number;
+	[x: symbol]: boolean;
+	foo: string;
+	bar: Date;
+	baz: boolean;
+}>(fooBar);
+
+declare function setFooBar(fooBar: FooBar): void;
+
+expectError(setFooBar({
+	[Symbol(42)]: 'life',
+	foo: 'foo',
+	bar: new Date(),
+	baz: true,
+}));

